### ISSUE TITLE
Cross-batch operations while using multi-gpu

### ISSUE THREAD

**Issue by wasiahmad** [pytorch/fairseq#2936]
Hi,

I need to execute a criterion to compute the loss value that requires interactions of the generated representations by a model. In the multi-GPU setting, mini-batches are constructed and the backward propagation is performed in each individual GPU. Usually, there is a function called `train_step` in a defined task that computes the loss and performs backpropagation, such as [[1]](https://github.com/pytorch/fairseq/blob/master/fairseq/tasks/translation_lev.py#L173).

But I need to compute loss based on the outputs of the entire batch (across all GPUs). What is the best way to do it?

**Comment by myleott** [comment]
Can you clarify a bit more about what your loss is? In particular, autograd is unlikely to work properly out of the box when computing loss terms over multiple GPUs.

You'll probably need to implement a custom torch.autograd.Function that does some kind of `torch.distributed.*` calls in the forward and backward. An example of this is VocabParallelCrossEntropy, which shards the output over the vocabulary and computes the loss across multiple (model parallel) workers: https://github.com/ngoyal2707/Megatron-LM/blob/adb23324c222aad0aad89308e70302d996a5eaeb/mpu/cross_entropy.py

**Comment by wasiahmad** [comment]
@myleott I have a question batch and a document batch and I need to perform ranking. I want to utilize the idea of [DPR](https://github.com/facebookresearch/DPR). So, for each question, all the instances in the document batch is a candidate. Now the critical part is, I am using 8 11GB GPUs with a total batch size of 64 (each GPU processes mini-batches with 8 examples). The final ranking objective needs to be executed over all the question and document representations accumulated across GPUs. But I do not know how to do this in Fairseq as I have my own criterion which gets individual mini-batches (coming from each GPU). 

My underlying model architecture is BART and I have my own task and criterion. In the task, I defined the `train_step`. The only concern is the loss that I use to backpropagate is not computed over the global batch, it is rather computed locally for each mini-batch.

I am not highly familiar with distributed stuff, I hope you can suggest an easy way to tackle my need.
